_posts/2016-01-26-variance-and-functors.md

```diff
@@ -0,0 +1,282 @@
+  new Functor[Read] {
+    def map[A, B](fa: Read[A])(f: A => B): Read[B] =
+      new Read[B] {
+        def read(s: String): Option[B] =
+          fa.read(s) match {
+            case None => None
+            case Some(a) => Some(f(a))
+          }
+      }
+  }
+```
+
+With that we can do useful things like
+
+```scala
+val circles: List[Circle] = List(Circle(..), Circle(..))
+val shapes: List[Shape] = listFunctor.map(circles)(circle => circle: Shape) // upcast
+
+val parseCircle: Read[Circle] = ...
+val parseShape: Read[Shape] = readFunctor.map(parseCircle)(circle => circle: Shape) // upcast
+```
+
+or more generally:
+
+```scala
+def upcast[F[_], A, B <: A](functor: Functor[F], fb: F[B]): F[A] =
+  functor.map(fb)(b => b: A)
+```
+
+`upcast`'s behavior does exactly what covariance does - given some supertype `A` (`Shape`) and a subtype `B` (`Circle`),
+we can mechanically (and safely) turn an `F[B]` into an `F[A]`. Put differently, anywhere we expect an `F[A]` we can provide
+an `F[B]`, i.e. covariance. For this reason, `Functor` is sometimes referred to in full as covariant functor.
+
+# Contravariance
+Contravariance flips the direction of the relationship in covariance - an `F[Shape]` is considered a
+subtype of `F[Circle]`. This seems strange - when I was first learning about variance I couldn't
+come up with a situation where this would make sense.
+
+If we have a `List[Shape]` we cannot safely treat it as a `List[Circle]` - doing so comes with all the usual
+warnings about downcasting. Similarly if we have a `Read[Shape]`, we cannot treat it as a `Read[Circle]` -
+we know how to parse a `Shape`, but we don't know how to parse any additional information `Circle` may need.
+
+## Show
+It appears fundamentally read-only types cannot be treated as contravariant. However, given that contravariance
+is covariance with the direction reversed, can we also reverse the idea of a read-only type? Instead of reading
+a value *from* a `String`, we can write a value *to* a `String`.
+
+```scala
+trait Show[-A] {
+  def show(a: A): String
+}
+```
+
+`Show` is the other side of `Read` - instead of going from a `String` to an `A`, we go from an `A` into
+a `String`. This reversal allows us to define contravariant behavior - if we are asked to provide a way
+to show a `Circle` (`Show[Circle]`), we can give instead a way to show just a `Shape`. This is a valid
+substitution because we can show a `Circle` by throwing away `Circle`-specific information and showing just
+the `Shape` bits. This means that `Show[Shape]` is a subtype of `Show[Circle]`, despite `Circle` being a
+subtype of `Shape`.
+
+In general, we can show (or write) a subtype if we know how to show a supertype by tossing away subtype-specific
+information (an upcast) and showing the remainder. Again, this means `Show[Supertype]` is substitutable, or a
+subtype of, `Show[Subtype]`.
+
+For similar reasons that read-only types can be made covariant, write-only types can be made contravariant.
```

Comment:
This may be a bit confusing. By "types" are you referring to the type class? We've shown that we can both "read" and "write" a `Shape` so "read-only" and "write-only" don't seem to apply to `Shape` itself.